module/version_reconfigure_test.go

```diff
@@ -0,0 +1,260 @@
+package module_test
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+	"strconv"
+	"testing"
+	"time"
+
+	"github.com/edaniels/golog"
+	"github.com/pkg/errors"
+	"go.viam.com/test"
+	goutils "go.viam.com/utils"
+	"go.viam.com/utils/pexec"
+	"go.viam.com/utils/rpc"
+
+	"go.viam.com/rdk/components/generic"
+	"go.viam.com/rdk/config"
+	"go.viam.com/rdk/robot"
+	"go.viam.com/rdk/robot/client"
+	"go.viam.com/rdk/testutils"
+	"go.viam.com/rdk/utils"
+)
+
+// TODO(pre-merge) refactor this somewhat-shared function into a test_helpers esque file.
+func connect(port string, logger golog.Logger) (robot.Robot, error) {
+	connectCtx, cancelConn := context.WithTimeout(context.Background(), time.Second*30)
+	defer cancelConn()
+	for {
+		dialCtx, dialCancel := context.WithTimeout(context.Background(), time.Millisecond*500)
+		rc, err := client.New(dialCtx, "localhost:"+port, logger,
+			client.WithDialOptions(rpc.WithForceDirectGRPC()),
+			client.WithDisableSessions(), // TODO(PRODUCT-343): add session support to modules
+		)
+		dialCancel()
+		if !errors.Is(err, context.DeadlineExceeded) {
+			return rc, err
+		}
+		select {
+		case <-connectCtx.Done():
+			return nil, connectCtx.Err()
+		default:
+		}
+	}
+}
+
+// TODO(pre-merge) refactor this somewhat-shared function into a test_helpers esque file.
+func modifyCfg(t *testing.T, cfgIn string, logger golog.Logger) (string, string, error) {
+	p, err := goutils.TryReserveRandomPort()
+	if err != nil {
+		return "", "", err
+	}
+	port := strconv.Itoa(p)
+
+	cfg, err := config.Read(context.Background(), cfgIn, logger)
+	if err != nil {
+		return "", "", err
+	}
+	cfg.Network.BindAddress = "localhost:" + port
+	output, err := json.Marshal(cfg)
+	if err != nil {
+		return "", "", err
+	}
+	file, err := os.CreateTemp(t.TempDir(), "viam-multiversionmodule-config-*")
+	if err != nil {
+		return "", "", err
+	}
+	cfgFilename := file.Name()
+	_, err = file.Write(output)
+	if err != nil {
+		return "", "", err
+	}
+	return cfgFilename, port, file.Close()
+}
+
+func TestValidationFailureDuringReconfiguration(t *testing.T) {
+	logger, logs := golog.NewObservedTestLogger(t)
+
+	// Use a valid config
+	cfgFilename, port, err := modifyCfg(t, utils.ResolveFile("module/multiversionmodule/module.json"), logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	serverPath, err := testutils.BuildTempModule(t, "web/cmd/server/")
+	test.That(t, err, test.ShouldBeNil)
+
+	server := pexec.NewManagedProcess(pexec.ProcessConfig{
```

Comment:
I realize @zaporter-work wrote most of this and is OOO at the moment, but I'm confused why we're testing this logic at such a high level. Why do we need to run the web server and mutate the config it's using? See other tests in `local_robot_test.go` for how we create new local robot with a certain config and then just call `Reconfigure` with a new config. I don't think we need to be testing sensitivity to file changes, do we?

Reply:
I do think there's something to be said for black box testing. Changing the tests did run into some Attributes/ConvertedAttributes finickiness. In doing the conversion, I didn't see anything special/wrong with how `local_robot_test.go` was doing module testing. And that removed some of the startup a process/get a connection/rewrite a json file complexity. So I think that's a better path here.